mllib/src/main/scala/org/apache/spark/mllib/clustering/KMeansModel.scala

```diff
@@ -39,6 +39,8 @@ import org.apache.spark.sql.{Row, SparkSession}
 class KMeansModel @Since("1.1.0") (@Since("1.0.0") val clusterCenters: Array[Vector])
   extends Saveable with Serializable with PMMLExportable {
 
+  private val clusterCentersWithNorm = clusterCenters.map(new VectorWithNorm(_))
```

Comment:
I believe `clusterCenters` will _still_ be a separate member, given the declaration. However, it will contain references to the same `Vector` objects. The vectors won't be duplicated, but some references will. This still seems like negligible overhead compared to the benefit; it's kilobytes of memory.